### PR TITLE
feature: Using Celery for APIs and periodic tasks

### PR DESCRIPTION
**Pull request by melvinebenezer**
### Summary
Made the changes for Celery Workers 
- APIs toxicity and Feature extraction
- Periodic Tasks like user streak update

### Some Issues
- HF some times returns 503 or 500 

### Test Evidence
Screenshot 2023-02-20 at 6 11 07 AM

### CODE REVIEW

**Comment by Nil-Andreu** [comment]
Just left some comments, do not want to overwhelm, just doubts.

If you have Discord @melvinebenezer, ping me please so we can have chat and could help on this PR (seems cool, so i am happy to contribute in it).

File: backend/oasst_backend/scheduled_tasks.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    # time.sleep(a)
+    logger.info(f"IN create task ... {a}")
+    return b + c
+
+
+async def useHFApi(text, url, model_name):
+    hugging_face_api: HuggingFaceAPI = HuggingFaceAPI(f"{url}/{model_name}")
+    # TODO: This is some wierd error that commenting the next line causes a 400 error
+    hugging_face_api.headers: Dict[str, str] = {"Authorization": f"Bearer {hugging_face_api.api_key}"}
+    logger.info(f"hf {hugging_face_api.headers}")
```
[the remark this answers is not in the document]
Comment (melvinebenezer) [reply backend/oasst_backend/scheduled_tasks.py]: Yes it was for debug. Removing it in the next update

File: backend/oasst_backend/scheduled_tasks.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+def create_task(a, b, c):
+    # time.sleep(a)
+    logger.info(f"IN create task ... {a}")
+    return b + c
+
+
+async def useHFApi(text, url, model_name):
+    hugging_face_api: HuggingFaceAPI = HuggingFaceAPI(f"{url}/{model_name}")
+    # TODO: This is some wierd error that commenting the next line causes a 400 error
+    hugging_face_api.headers: Dict[str, str] = {"Authorization": f"Bearer {hugging_face_api.api_key}"}
```
[the remark this answers is not in the document]
Comment (melvinebenezer) [reply backend/oasst_backend/scheduled_tasks.py:44]: This has been the current design choice. Yes, we can pass it as an environment variable. Probably as a separate PR

File: backend/oasst_backend/scheduled_tasks.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+result = load_dotenv(".env")
+logger.info(f"LoadEnv result {result}")
+
+app = Celery(__name__)
+app.conf.broker_url = os.environ.get("CELERY_BROKER_URL")
+app.conf.result_backend = os.environ.get("CELERY_RESULT_BACKEND")
+logger.info(f"celery.conf.broker_url {app.conf.broker_url}, app.conf.result_backend{app.conf.result_backend}")
+startup_time: datetime = utcnow()
+
+# Load task modules from all registered Django applications.
```
[the remark this answers is not in the document]
Comment (melvinebenezer) [reply backend/oasst_backend/scheduled_tasks.py]: Thanks, Removed it

File: backend/oasst_backend/scheduled_tasks.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        print(f"Could not compute toxicity for text reply to {message_id=} with {text=} by.error {str(e)}")
+
+
+@app.task(name="hf_feature_extraction")
+def hf_feature_extraction(text, message_id, api_client):
+    try:
+        with default_session_factory() as session:
+            model_name: str = HfEmbeddingModel.MINILM.value
+            url: str = HfUrl.HUGGINGFACE_FEATURE_EXTRACTION.value
+            embedding = async_to_sync(useHFApi)(text=text, url=url, model_name=model_name)
```
[the remark this answers is not in the document]
Comment (melvinebenezer) [reply backend/oasst_backend/scheduled_tasks.py:77]: Celery requires task functions to be synchronous. 

AsyncToSync lets a synchronous subthread stop and wait while the async function is called on the main thread's event loop, and then control is returned to the thread when the async function is finished.

**Comment by melvinebenezer** [comment]
@Nil-Andreu  thanks for the review comments. Have made the changes. open to any suggestions

File: backend/start_worker.sh [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
```
Comment (andreaskoepf) [inline backend/start_worker.sh:1]: run scripts are normally stored in `/scripts/backend_development`

File: backend/oasst_backend/tree_manager.py [hunk 6]
```diff
@@ -693,36 +695,51 @@ async def handle_interaction(self, interaction: protocol_schema.AnyInteraction)
...
-                        )
-                        embedding = await hugging_face_api.post(interaction.text)
-                        pr.insert_message_embedding(
-                            message_id=message.id, model=HfEmbeddingModel.MINILM.value, embedding=embedding
-                        )
+                        hf_feature_extraction.delay(interaction.text, message.id, pr.api_client.dict())
+                        logger.debug("Extract Embedding")
                     except OasstError:
-                        logger.error(
+                        logger.debug(
```
Comment (andreaskoepf) [inline backend/oasst_backend/tree_manager.py:701]: why is this no longer logged as error?

File: backend/main.py [hunk 7]
```diff
@@ -284,44 +288,46 @@ def update_leader_board_total(session: Session) -> None:
 
 
 @app.on_event("startup")
-@repeat_every(seconds=60 * 60 * settings.USER_STREAK_UPDATE_INTERVAL, wait_first=False)
-def update_user_streak(session: Session) -> None:
+@repeat_every(seconds=3600, wait_first=False)
+def update_user_streak() -> None:
```
Comment (andreaskoepf) [inline backend/main.py:292]: why is update_user_streak() now still part of main?

File: backend/oasst_backend/scheduled_tasks.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+from loguru import logger
+from oasst_backend.models import ApiClient
+from oasst_backend.prompt_repository import PromptRepository
+from oasst_backend.user_repository import User
+from oasst_backend.utils.database_utils import default_session_factory
+from oasst_backend.utils.hugging_face import HfClassificationModel, HfEmbeddingModel, HfUrl, HuggingFaceAPI
+from oasst_shared.utils import utcnow
+from sqlmodel import select
+
+result = load_dotenv(".env")
```
Comment (andreaskoepf) [inline backend/oasst_backend/scheduled_tasks.py:19]: This code in global scope looks not good. why is it here? Is this file an entry-point?

File: backend/oasst_backend/tree_manager.py [hunk 9]
```diff
@@ -693,36 +695,51 @@ async def handle_interaction(self, interaction: protocol_schema.AnyInteraction)
...
                             f"Could not fetch embbeddings for text reply to {interaction.message_id=} with {interaction.text=} by {interaction.user=}."
                         )
 
+                    # try:
+                    #     hugging_face_api = HuggingFaceAPI(
+                    #         f"{HfUrl.HUGGINGFACE_FEATURE_EXTRACTION.value}/{HfEmbeddingModel.MINILM.value}"
+                    #     )
+                    #     embedding = await hugging_face_api.post(interaction.text)
+                    #     pr.insert_message_embedding(
```
Comment (andreaskoepf) [inline backend/oasst_backend/tree_manager.py:710]: will the comments be removed?

File: docker/Dockerfile.backendworker [hunk 10]
```diff
@@ -0,0 +1,28 @@
+FROM python:3.10
```
Comment (andreaskoepf) [inline docker/Dockerfile.backendworker:1]: I suggest to name this file to `Dockerfile.backend-worker` ...

File: backend/oasst_backend/tree_manager.py [hunk 11]
```diff
@@ -22,13 +22,15 @@
...
+from oasst_backend.scheduled_tasks import hf_feature_extraction, toxicity
 from oasst_backend.utils.database_utils import (
     CommitMode,
     async_managed_tx_method,
     managed_tx_function,
     managed_tx_method,
 )
-from oasst_backend.utils.hugging_face import HfClassificationModel, HfEmbeddingModel, HfUrl, HuggingFaceAPI
+
+# from oasst_backend.utils.hugging_face import HfClassificationModel, HfEmbeddingModel, HfUrl, HuggingFaceAPI
```
Comment (andreaskoepf) [inline backend/oasst_backend/tree_manager.py:33]: remove?

File: backend/main.py [hunk 12]
```diff
@@ -137,6 +139,8 @@ async def http_callback(request: fastapi.Request, response: fastapi.Response, pe
     @app.on_event("startup")
     @managed_tx_function(auto_commit=CommitMode.COMMIT)
     def create_seed_data(session: Session):
+        # logger.debug("create_seed_data...")
```
Comment (andreaskoepf) [inline backend/main.py:142]: remove?

File: backend/oasst_backend/scheduled_tasks.py (hunk 8 above)
Comment (melvinebenezer) [reply backend/oasst_backend/scheduled_tasks.py]: Yes this is the entry function for the Celery Worker process

**Review by andreaskoepf — APPROVED** [review]
Thanks a lot .. also fro your patience!

**Pull request merged**